### PR TITLE
raptor: improve performances in foot paths

### PR DESCRIPTION
**Pull request by TeXitoi**
Only raptor, for 1000 trips in seconds:
- israel: 327.37 -> 179.25 => 45% faster
- idf: 123.4 -> 78.12 => 37% faster
- fr-pdl: 241.91 -> 122.93 => 49% faster

### CODE REVIEW

File: source/routing/raptor.cpp [hunk 1]
```diff
@@ -498,6 +512,25 @@ void RAPTOR::set_valid_jp_and_jpp(uint32_t date, const std::vector<std::string>
...
+    // propagate the invalid jp in their jpp
+    for (type::idx_t jp_idx = 0; jp_idx < valid_journey_patterns.size(); ++jp_idx) {
+        if (valid_journey_patterns[jp_idx]) { continue; }
+        const auto* jp = data.pt_data->journey_patterns[jp_idx];
+        for (const auto* jpp: jp->journey_pattern_point_list) {
+            valid_journey_pattern_points.set(jpp->idx, false);
+        }
+    }
+
+    jpps_from_sp = data.dataRaptor->jpps_from_sp;
```
Comment (antoine-de) [inline source/routing/raptor.cpp:540]: lache un com'!

File: source/routing/raptor_path.cpp [hunk 2]
```diff
@@ -52,14 +52,14 @@ makePathes(const std::vector<std::pair<type::idx_t, navitia::time_duration> > &d
...
-get_current_stidx_gap(size_t count, type::idx_t jpp_idx, const std::vector<Labels> &labels,
+get_current_stidx_gap(size_t count, JppIdx jpp_idx, const std::vector<Labels> &labels,
                       const type::AccessibiliteParams & accessibilite_params, bool clockwise,  const navitia::type::Data &data, bool disruption_active) {
     const auto& ls = labels[count];
     if(ls.pt_is_initialized(jpp_idx)) {
         const auto& dt_pt = ls.dt_pt(jpp_idx);
         const auto date = DateTimeUtils::date(dt_pt);
         const auto hour = DateTimeUtils::hour(dt_pt);
-        const type::JourneyPatternPoint* jpp = data.pt_data->journey_pattern_points[jpp_idx];
+        const type::JourneyPatternPoint* jpp = data.pt_data->journey_pattern_points[jpp_idx.val];
```
Comment (antoine-de) [inline source/routing/raptor_path.cpp]: can't you do a 'get_jpp' ?

File: source/routing/raptor_path_defs.h [hunk 3]
```diff
@@ -7,11 +7,11 @@
 namespace navitia { namespace routing {
 
 template<typename Visitor>
-void handle_connection(const size_t countb, const navitia::type::idx_t current_jpp_idx, Visitor& v,
+void handle_connection(const size_t countb, const JppIdx current_jpp_idx, Visitor& v,
                        bool clockwise, const RAPTOR &raptor_) {
-    const auto& departure = raptor_.data.pt_data->journey_pattern_points[current_jpp_idx]->stop_point;
+    const auto& departure = raptor_.data.pt_data->journey_pattern_points[current_jpp_idx.val]->stop_point;
```
Comment (antoine-de) [inline source/routing/raptor_path_defs.h]: get_jpp no ?

File: source/routing/raptor_path_defs.h [hunk 4]
```diff
@@ -7,11 +7,11 @@
...
     const auto& boarding_jpp_idx = raptor_.labels[countb].boarding_jpp_transfer(current_jpp_idx);
-    const auto& destination_jpp = raptor_.data.pt_data->journey_pattern_points[boarding_jpp_idx];
+    const auto& destination_jpp = raptor_.data.pt_data->journey_pattern_points[boarding_jpp_idx.val];
```
Comment (antoine-de) [inline source/routing/raptor_path_defs.h]: get_jpp no ?

File: source/routing/raptor_solutions.cpp [hunk 5]
```diff
@@ -59,14 +59,14 @@ get_solutions(const std::vector<std::pair<type::idx_t, navitia::time_duration> >
 
 
 Solutions
-get_solutions(const std::vector<std::pair<type::idx_t, navitia::time_duration> > &departs,
+get_solutions(const std::vector<std::pair<SpIdx, navitia::time_duration> > &departs,
               const DateTime &dep, bool clockwise, const type::Data & data, bool) {
     Solutions result;
     for(auto dep_dist : departs) {
-        for(auto journey_pattern : data.pt_data->stop_points[dep_dist.first]->journey_pattern_point_list) {
+        for(const auto* jpp: data.pt_data->stop_points[dep_dist.first.val]->journey_pattern_point_list) {
```
Comment (antoine-de) [inline source/routing/raptor_solutions.cpp]: get_sp no ?

File: source/routing/raptor_utils.h [hunk 6]
```diff
@@ -100,19 +105,20 @@ struct Labels {
...
+    IdxMap<type::JourneyPatternPoint, DateTime> dt_transfers;
     // jpp used to reach this label with public transport
-    std::vector<navitia::type::idx_t> boarding_jpp_pts;
+    IdxMap<type::JourneyPatternPoint, JppIdx> boarding_jpp_pts;
     // jpp used to reach this label with a transfer
-    std::vector<navitia::type::idx_t> boarding_jpp_transfers;
+    IdxMap<type::JourneyPatternPoint, JppIdx> boarding_jpp_transfers;
 };
 
+// TODO: cleanup
```
Comment (antoine-de) [inline source/routing/raptor_utils.h]: remove those so no ?

File: source/routing/raptor_utils.h [hunk 7]
```diff
@@ -124,27 +130,27 @@ inline void memset32(T*buf, uint n, T c)
...
+    IdxMap<type::JourneyPatternPoint, navitia::time_duration> jpp_idx_duration;
     DateTime best_now;
-    type::idx_t best_now_jpp_idx;
+    JppIdx best_now_jpp_idx;
     size_t count;
 
-    void add_destination(const type::JourneyPatternPoint* jpp, const time_duration duration_to_dest) {
-        jpp_idx_duration[jpp->idx] = duration_to_dest; //AD, check if there are some rounding problems
+    void add_destination(const JppIdx jpp, const time_duration duration_to_dest) {
+        jpp_idx_duration[jpp] = duration_to_dest; //AD, check if there are some rounding problems
```
Comment (antoine-de) [inline source/routing/raptor_utils.h]: I think we can remove this comment :)

File: source/routing/tests/raptor_test.cpp [hunk 8]
```diff
@@ -1333,13 +1333,14 @@ BOOST_AUTO_TEST_CASE(less_fallback) {
     RAPTOR raptor(*(b.data));
     type::PT_Data & d = *b.data->pt_data;
 
-    std::vector<std::pair<type::idx_t, navitia::time_duration>> departs = {
-        {d.stop_areas_map["stop1"]->stop_point_list.front()->journey_pattern_point_list.front()->idx, navitia::seconds(0)}
+    std::vector<std::pair<SpIdx, navitia::time_duration>> departs = {
+        {SpIdx(*d.stop_areas_map["stop1"]->stop_point_list.front()), navitia::seconds(0)}
+    };
+    std::vector<std::pair<SpIdx, navitia::time_duration>> destinations = {
+        {SpIdx(*d.stop_areas_map["stop1"]->stop_point_list.front()), navitia::seconds(560)},
```
Comment (antoine-de) [inline source/routing/tests/raptor_test.cpp]: what, you don't use the beautiful `560_s` syntax ??!

**Comment by antoine-de** [comment]
OK for me, can you also check @l-vincent-l  ?

File: source/routing/raptor_path.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/routing/raptor_path.cpp]: I can't, I don't have the RAPTOR object.

File: source/routing/raptor_solutions.cpp (hunk 5 above)
Comment (TeXitoi) [inline source/routing/raptor_solutions.cpp]: I can't, I don't have a RAPTOR object.

File: source/routing/raptor.cpp [hunk 9]
```diff
@@ -498,6 +514,29 @@ void RAPTOR::set_valid_jp_and_jpp(uint32_t date, const std::vector<std::string>
         }
     }
 
+    // filter accessibility
```
Comment (l-vincent-l) [inline source/routing/raptor.cpp:517]: you may test if there are some filtered parameters in accessibilite_params first for optimization's sake.

**Comment by antoine-de** [comment]
Nice work on adding types!

can you put figures about the performance gain ?

**Pull request merged**